google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -0,0 +1,568 @@
+    bulkWriter.close().get();
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result.get().getUpdateTime());
+  }
+
+  @Test
+  public void hasUpdateMethod() throws Exception {
+    RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(update(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(2));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock);
+
+    ApiFuture<WriteResult> result = bulkWriter.update(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close().get();
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result.get().getUpdateTime());
+  }
+
+  @Test
+  public void hasDeleteMethod() throws Exception {
+    RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(batchWrite(delete("coll/doc1")), successResponse(2));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock);
+
+    ApiFuture<WriteResult> result = bulkWriter.delete(doc1);
+    bulkWriter.close().get();
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result.get().getUpdateTime());
+  }
+
+  @Test
+  public void hasCreateMethod() throws Exception {
+    RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(create(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(2));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock);
+
+    ApiFuture<WriteResult> result = bulkWriter.create(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close().get();
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result.get().getUpdateTime());
+  }
+
+  @Test
+  public void surfacesErrors() throws Exception {
+    RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                failResponse());
```

Comment:
Should we pass the error code to `failResponse`? 

Nit: I think the name should be `failedResponse`, since it returns a failed response.